src/transforms/sample/transform.rs

```diff
@@ -12,24 +12,112 @@ use crate::{
 use vector_lib::lookup::lookup_v2::OptionalValuePath;
 use vector_lib::lookup::OwnedTargetPath;
 
+/// Exists only for backwards compatability purposes so that the value of sample_rate_key is
+/// consistent after the internal implementation of the Sample class was modified to work in terms
+/// of percentages
+#[derive(Clone, Debug)]
+pub enum SampleMode {
+    Rate {
+        rate: u64,
+        counters: HashMap<Option<String>, u64>,
+    },
+    Ratio {
+        ratio: f64,
+        values: HashMap<Option<String>, f64>,
+        hash_ratio_threshold: u64,
+    },
+}
+
+impl SampleMode {
+    pub fn new_rate(rate: u64) -> Self {
+        Self::Rate {
+            rate,
+            counters: HashMap::default(),
+        }
+    }
+
+    pub fn new_ratio(ratio: f64) -> Self {
+        Self::Ratio {
+            ratio,
+            values: HashMap::default(),
+            // Supports the 'key_field' option, assuming an equal distribution of values for a given
+            // field, hashing its contents this component should output events according to the
+            // configured ratio.
+            //
+            // To do this convert the hash to a number between 0 and 1 and compare to the ratio, to
+            // address issues with precision here the ratio is expanded to meet the width of the type of
+            // the hash.
```

Comment:
comment nit: You're actually not converting the hash to anything, the ratio is scaled up to match the range of the possible hash value.

Reply:
Maybe the comment is confusing but I'm stating how that is the solution can be seen and the inverse instead is performed. I can clarify it.

Reply:
Ah, okay, fair enough.